### ISSUE TITLE
Error inserting large values into varbinary(max) on SQL Server 2008.

### ISSUE THREAD

**Issue by GoogleCodeExporter** [maximemaher/pyodbc#277]
```
What steps will reproduce the problem?

When I try to insert a large value into a varbinary(max) column on SQL Server 
2008 it is producing an error "The text, ntext, or image pointer value 
conflicts with the column name specified. (7125)"  Based on this MSDN article 
http://msdn.microsoft.com/en-us/library/ms130896.aspx this problem seems to be 
tied to the data bindings that the ODBC driver is invoking.

Here is a sample DDL:
use TestDb 
Go

create table blobTest (
    id int identity primary key,
    dataDesc varchar(200),
    imageData varbinary(max)
    )

And sample code that generates the error:
import pyodbc, pickle
from PIL import Image
from StringIO import StringIO

#Create the SQL connection
sqlConnStr = ('DRIVER={SQL Server};Server=(local);Database=TestDb;'+
            'Trusted_Connection=YES')

sqlConn = pyodbc.connect(sqlConnStr)

#Create a image to store and prepare it for starge
img = Image.new('RGB', #mode
                (1024, 768), #size
                (0, 0, 255) #color, this is dark blue
                )
imgBuffer = StringIO()
img.save(imgBuffer, "BMP")
img.save('testblue.bmp')

#Store the image in the database
curs = sqlConn.cursor()
sql = """
    insert into blobTest (dataDesc, imageData) 
    values (?, ?)"""
curs.execute(sql, ('Blue Image', pyodbc.Binary(imgBuffer.getvalue())))
sqlConn.commit()

This code works quite nicely if the image is made smaller before trying to 
insert it.



What is the expected output? What do you see instead?

Receiving a "The text, ntext, or image pointer value conflicts with the column 
name specified. (7125)"   error.

What version of the product are you using? On what operating system?
This is version 3.0.2 on Python 2.7.1 running on Windows 7


Please provide any additional information below.


```

Original issue reported on code.google.com by `timothyawiseman@gmail.com` on 17 Jul 2012 at 9:54

**Comment by GoogleCodeExporter** [comment]
```
Exactly how big, in bytes, is the value that fails?

I've updated the SQL Server unittests (tests2/sqlservertests) to include 'max' 
tests for varchar, nvarchar, and binary and to include 5MB and 50MB inserts, 
which work.

```

Original comment by `mkleehammer` on 29 Sep 2012 at 3:06
- Changed state: **Hold**

**Comment by GoogleCodeExporter** [comment]
Original comment by `mkleehammer` on 29 Sep 2012 at 3:10
- Added labels: **Driver-SQLServer**

**Comment by GoogleCodeExporter** [comment]
```
The imbBuffer.len property returns 2359350 which matches the 2,305KB size of 
the file if I save it as a file.  I also receive the error at 640x480 (921654 
bytes)

If I reduce the size of the image to 320x420 which takes 230454 bytes the 
script executes perfectly and SQL Server confirms that it stores the varbinary 
data.

I also tested this on a completely separate machine with SQL Server 2012 
Express with the same results.
```

Original comment by `timothyawiseman@gmail.com` on 29 Sep 2012 at 6:22

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
And this is where I realize how I have been overlooking something obvious and 
apologize for wasting your time.  I work with numerous machines with several 
versions of SQL Server, and it never occurred to me while I was troubleshooting 
this to check that I was using the right driver for the version I was using.  I 
failed to notice this even though I made certain to test this on more than one 
machine and more than one install of SQL Server 2008 (and later 2012) before 
filing the bug report.

In short, when I switch to DRIVER={SQL Server Native Client 10.0} instead of 
DRIVER={SQL Server} in the connection string, the error vanishes and works 
properly.

I finally figured this out after playing with and then starting to actually 
read through the SQL Server Unittests you provided.  

Thanks again.
```

Original comment by `timothyawiseman@gmail.com` on 10 Oct 2012 at 10:10

**Comment by GoogleCodeExporter** [comment]
```
@ Timothy A Wiseman: I experienced a very similar issue but with inserting 
Geometry data into a varchar(max) field (in a tmp table so it could be inserted 
into a Geometry data type field). Using the Sql Server driver, it would throw 
the error you described but once I switched to the Native Client 11.0 driver, 
it was successful. Thanks for posting this.
```

Original comment by `gisch...@gmail.com` on 31 Jul 2015 at 7:44

**Comment by bkline** [comment]
I see the behavior originally reported with both the deprecated SQL Server Native Client NN.0 drivers as well as the recommended ODBC NN drivers, when running under Python 3.6. I can't reproduce the problem running under Python 2.7, no matter how large the value is. The threshold for the failure is 8001 bytes. Values 8000 bytes or smaller don't have a problem with either version of Python.

```
from argparse import ArgumentParser
from pyodbc import connect, Binary

parser = ArgumentParser()
parser.add_argument("--server", "-s", required=True)
parser.add_argument("--database", "-d", required=True)
parser.add_argument("--user", "-u", required=True)
parser.add_argument("--password", "-p", required=True)
parser.add_argument("--num_bytes", "-n", type=int, default=8001)
opts = parser.parse_args()
parms = dict(
    Driver="{ODBC Driver 13 for SQL Server}",
    Server=opts.server,
    Database=opts.database,
    Uid=opts.user,
    Pwd=opts.password
)
parms["Driver"] = "{SQL Server Native Client 11.0}"
b = b'x' * opts.num_bytes
connection_string = ";".join(["{}={}".format(*p) for p in parms.items()])
connection = connect(connection_string)
cursor = connection.cursor()
cursor.execute("CREATE TABLE #t (b VARBINARY(MAX))")
cursor.execute("INSERT INTO #t VALUES(?)", (b,))
cursor.execute("SELECT b FROM #t")
b2 = cursor.fetchone().b
if b == b2:
    print("OK")
else:
    print("mismatch!")
```

With Python 2.7, or with a value of 8000 or less for the --num_bytes option, the output is `OK`.

With Python 3.6 and the default value of 8001 for the --num_bytes option, the output is:

```
Traceback (most recent call last):
  File "issue-277-repro.py", line 24, in <module>
    cursor.execute("INSERT INTO #t VALUES(?)", (b,))
pyodbc.Error: ('HY104', '[HY104] [Microsoft][SQL Server Native Client 11.0]Invalid precision value (0) (SQLBindParameter)')
```

(Commenting out the `parms["Driver"] = ...` line gets the same outcome, except for the driver string displayed in the exception output.)

I'm running pyodbc 4.0.22 under Python 3.6.2 (v3.6.2:5fd33b5, Jul  8 2017, 04:57:36) [MSC v.1900 64 bit (AMD64)] on Microsoft Windows [Version 6.1.7601] against Microsoft SQL Server 2016 (SP1-GDR) (KB4057118) - 13.0.4210.6 (X64)   Dec 22 2017 15:00:27   Copyright (c) Microsoft Corporation  Developer Edition (64-bit) on Windows Server 2012 R2 Standard 6.3 <X64> (Build 9600: ) (Hypervisor).

**Comment by ThomIves** [comment]
Confirmation that using the driver string below fixed my exactly similar issue - thanks!

self.conn = pyodbc.connect(
                driver='{**SQL Server Native Client 11.0**}', server='TM1',
                database=self.database,
                trusted_connection='yes')

**Comment by bkline** [comment]
Actually, I don't think that's what fixed your problem, as that's the same driver as one of the ones I reported as failing. It appears that the developers of `pyodbc` fixed the problem, but just didn't report the fix for this ticket. (Perhaps they were fixing something else and they didn't realize they had fixed this bug, too.) Here are the results of running a version of the script I posted above back in 2018, adapted to use all three drivers installed on our Windows servers in a loop, run first with the version of `pyodbc` I was using when I reported the problem, and then with the current version of `pyodbc`. In this version I'm catching and printing the exception myself instead of letting it crash the script.

```
PS H:\src\python> ./pyodbc-issue277.py
pyodbc version 4.0.22
('HY104', '[HY104] [Microsoft][ODBC Driver 13 for SQL Server]Invalid precision value (0) (SQLBindParameter)')
('HY104', '[HY104] [Microsoft][ODBC Driver 17 for SQL Server]Invalid precision value (0) (SQLBindParameter)')
('HY104', '[HY104] [Microsoft][SQL Server Native Client 11.0]Invalid precision value (0) (SQLBindParameter)')
PS H:\src\python> pip install -U pyodbc
Collecting pyodbc
  Using cached pyodbc-4.0.30-cp38-cp38-win_amd64.whl (68 kB)
Installing collected packages: pyodbc
  Attempting uninstall: pyodbc
    Found existing installation: pyodbc 4.0.22
    Uninstalling pyodbc-4.0.22:
      Successfully uninstalled pyodbc-4.0.22
Successfully installed pyodbc-4.0.30
PS H:\src\python> ./pyodbc-issue277.py
pyodbc version 4.0.30
{ODBC Driver 13 for SQL Server} OK
{ODBC Driver 17 for SQL Server} OK
{SQL Server Native Client 11.0} OK
```

I haven't done the investigation to find out in which release of `pyodbc` the fix first landed, but we do have a machine still running 4.0.27 on which the tests pass.

**Comment by bkline** [comment]
I just noticed that this isn't the same repository as the one to which I reported the bug. I have no idea what the relationship between this repository and https://github.com/mkleehammer/pyodbc is, but that project has this ticket closed with information about when the bug was fixed. Very confusing to have a second repository without any indication that it's an out-of-date fork of the original. 👎
